### ISSUE TITLE
HttpClient.post body param is not sent correctly when using 'application/x-www-form-urlencoded' content type request header

### ISSUE THREAD

**Issue by TitaneBoy** [angular/angular#19535]
## I'm submitting a...

<pre><code>
[x] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
</code></pre>

## Current behavior

When sending http post request (using HttpClient) with body data (for example a json object), it looks that Angular send the json object as a "key" of another object.

For example, when doing `this.httpClient.post('https://localhost/login', JSON.stringify({username, password})`, if username is equal to (for example) 'user' and password is equal to 'pass', the request object in the server side will contain a body value like this: `{{"username": "user", "password": "pass"}:""}`

Here is a screenshot of what we can see as the value of "body" in request object, in server-side, 

![image](https://user-images.githubusercontent.com/15614466/31149185-15916500-a85d-11e7-9548-15066c67a1f8.png)

As you can see, the body object I'm trying to send with HttpClient.post seems to be the "key" of another Object wrapped around it, with an empty value..But it should be an object

## Expected behavior

Angular should send body object as an object, and not as the key of another object

## Minimal reproduction of the problem with instructions

Here is the code of my authentication service:

```
import { HttpClient, HttpHeaders, HttpResponse } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Response } from '@angular/http';
import 'rxjs/add/operator/map';
import { Observable } from 'rxjs/Observable';

@Injectable()
export class AuthenticationService {
    public token: string;

    constructor(private httpClient: HttpClient) {
        // set token if saved in local storage
        const currentUser = JSON.parse(localStorage.getItem('currentUser'));
        this.token = currentUser && currentUser.token;
    }

    login(username: string, password: string): Observable<boolean> {
        return this.httpClient.post('https://localhost/login', JSON.stringify({username, password}), {
          headers: new HttpHeaders().set('Content-Type', 'application/x-www-form-urlencoded'),
        }).map((response: Response) => {
          // login successful if there's a jwt token in the response
          const token = response.json() && response.json().token;
          if (token) {
              // set token property
              this.token = token;

              // store username and jwt token in local storage to keep user logged in between page refreshes
              localStorage.setItem('currentUser', JSON.stringify({ username, token }));

              // return true to indicate successful login
              return true;
          } else {
              // return false to indicate failed login
              return false;
          }
        });
    }

    logout(): void {
        // clear token remove user from local storage to log user out
        this.token = null;
        localStorage.removeItem('currentUser');
    }
}

```

In the server side, I'm using Node.js (v8.4.0) with Express (v4.16.0)

## Environment

<pre><code>
Angular version: 4.4.4

Browser:
- [X] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 8.4.0  
- Platform:  Windows 

Others:

</code></pre>

Thank you in advance for your help

**Comment by jonathanp** [comment]
Why do you `JSON.stringify` the object?

**Comment by TitaneBoy** [comment]
Well...It is an attempt to find a solution like any other. Using `JSON.stringify` or not will have the same behavior/issue. But I understand your point ;-)

**Comment by jonathanp** [comment]
`this.httpClient.post('https://localhost/login', JSON.stringify({username, password})` and `this.httpClient.post('https://localhost/login', {username, password})` will *not* have the same behaviour. Are you sure that the latter won't solve your problem?

**Comment by TitaneBoy** [comment]
Hi @jonathanp .. Yes..Unfortunetly, I have the same behavior in the server side..I've just tested it.

Here is a printscreen of the post function in client-side:

![image](https://user-images.githubusercontent.com/15614466/31180099-b3589798-a8eb-11e7-87cf-626b0ee08bd2.png)

And here is a printscreen of what I receive in the request object

![image](https://user-images.githubusercontent.com/15614466/31180196-fa894a04-a8eb-11e7-8b01-dd1c5d40fae2.png)

As you can see, the body data(` {username: "user", password: "pass"}`) I am sending with the HttpClient.post function becomes an Object key with empty value (` {"username": "user", "password": "pass"}:""`) in the request object of the server-side

**Comment by TitaneBoy** [comment]
I have captured with Wireshark software the data sended and it confirms the behavior I've described before:

![image](https://user-images.githubusercontent.com/15614466/31183367-5bd71856-a8f4-11e7-9bd6-ae35adfabd1e.png)

**Comment by niikoo** [comment]
I have the same problem with Angular 4.4.4

**Comment by MishoShalamberidze** [comment]
I have the same problem with Angular 4.4.4 :(

**Comment by TitaneBoy** [comment]
Hi Angular teams...Do you have any news / confirmations about this issue ?

Thanks you for your answer

**Comment by fk-mbc** [comment]
Hi TitaneBoy,

I think I had the same problem as you. Let me short explain. For the backend I use a nodejs express REST API. At first I run in the problem, that the body from the post requests was empty when I send a 
request via POSTMAN. So I found a solution and set the body parser for "application/x-www-form-urlencoded" to `app.use(bodyParser.urlencoded({ extended: true }));` in my node js app .

But now with Angular httpClient and `headers: new HttpHeaders().set('Content-Type', 'application/x-www-form-urlencoded')`, like you used as well, the request body is send in this ugly way. (Could be, that it is an bug, because from Postman it looks like expected).

**Solution for your Problem**: You need to send the data as json data to your backend by using "application/json" (which is the default 'Content-Type'). Then it could be, that you will run in the same behavior, that your backend app will have an empty response body. To fix this, you have to determine, that you want to use `app.use(bodyParser.json());` instead of `bodyParser.urlencoded`. That's all. 

When you want to send the request via Postman, you could use `raw` in the body section. Select JSON and pass in an json object.

I hope it will help you as well.

**Comment by TitaneBoy** [comment]
Hi @fk-mbc 

Thank you for you great detailed post.
In my case, my nodejs server side is already configured with the same configuration you described before ( `app.use(bodyParser.urlencoded({ extended: true }));`). I am using `bodyParser.json()` as well

![image](https://user-images.githubusercontent.com/15614466/31910248-dd0e329a-b80a-11e7-872f-b0e90abb719e.png)


And as you said, with "application/x-www-form-urlencoded" in request headers, I have **NO** issue when using POSTMAN and I receive correctly in the server-side the data sended in the body request. 

![image](https://user-images.githubusercontent.com/15614466/31910421-6d51482e-b80b-11e7-9b42-2c3f1249eef4.png)

Here is the data I receive in server-side:
![image](https://user-images.githubusercontent.com/15614466/31910475-96be8e06-b80b-11e7-81f2-beeffc947b89.png)

But when using HttpClient.post from Angular, the body data received in the server-side are ugly and as described before in this issue.

So..The only thing I can confirm now is that it works well with POSTMAN but not with HttpClient

**Comment by fk-mbc** [comment]
@TitaneBoy did you try to change `headers: new HttpHeaders().set('Content-Type', 'application/x-www-form-urlencoded')` to `headers: new HttpHeaders().set('Content-Type', 'application/json')` in your app on line 19? Because httpClient with header application/json works well.

**Comment by TitaneBoy** [comment]
@fk-mbc I've tried to change headers as you said. But I get some other problems...

![image](https://user-images.githubusercontent.com/15614466/31911888-1c634aa2-b810-11e7-976a-8281454b33c9.png)

So I decided to add options for my "login" route 

![image](https://user-images.githubusercontent.com/15614466/31911939-4669fe90-b810-11e7-8b3e-9414b28e66aa.png)

But even with this, I got another issue:

![image](https://user-images.githubusercontent.com/15614466/31911960-5c567814-b810-11e7-897a-9c5940d274f6.png)

So for now, your suggestion is not working for me and I'm trying to understand why..But I think is a complete different issue from what I've described in the beginning  of this post. I guess using HttpClient.post with 'application/x-www-form-urlencoded' should work as well as if it is used with 'application/json' content type header

**Comment by TitaneBoy** [comment]
@fk-mbc 
With Postman, I have no problem when using 'application/json' as well as 'application/x-www-form-urlencoded'. But I have the problems described in my last post when using Angular.

Could you share please with me the configuration of your route (in node) ? Did you add (like me) an ".options" for your route like you can see in my last screenshot ? Maybe I'm missing something...What is your node and express version ?

Thank you for your answer.

**Comment by TitaneBoy** [comment]
@fk-mbc 
I've just removed the "next()" instruction in my ".options" function and It works well now..

Web-client page

![image](https://user-images.githubusercontent.com/15614466/31914030-2169306e-b817-11e7-9204-9817bb93c892.png)

Client-Side

![image](https://user-images.githubusercontent.com/15614466/31914186-b5c71514-b817-11e7-9477-4e025efdb9a7.png)

Server-side

![image](https://user-images.githubusercontent.com/15614466/31914062-40a20910-b817-11e7-9d9e-20ef41a6b346.png)





It can be a workaround but I guess it is not working at all with 'application/x-www-form-urlencoded' content type header. I will rename the title of this issue to make focus of this aspect

**Comment by TitaneBoy** [comment]
@fk-mbc ...Thank you again for your suggestions and help :-)

**Comment by alxhub** [comment]
Hello @TitaneBoy,

I agree with some of the other commenters here - you cannot use `application/x-www-form-urlencoded` as your `Content-Type` and send a JSON encoded body. This is essentially lying to the server, and telling it to parse the body using a different format from the one it's actually encoded in.

What's happening is that your server-side framework is interpreting the JSON string as one long parameter name, with an empty value. Either encode your body in `application/x-www-form-urlencoded` format (pass an `HttpParams` instance as the body), or use `application/json`.

**Comment by fk-mbc** [comment]
@TitaneBoy I am  glad to hear, that you could solve the problem with my suggestions. :-)

**Comment by nischi** [comment]
If i do it like that, it works fine for me.

`
const body = new HttpParams()
    .set('client_id', Settings.client_id)
    .set('client_secret', Settings.client_secret)
    .set('grant_type', 'password')
    .set('scope', Settings.scope)
    .set('username', auth.username)
    .set('password', auth.password);

    return this.http.post(Settings.iam_service_base_url + 'connect/token', body.toString(), {
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      })
`

**Comment by vishal-bluesys** [comment]
@TitaneBoy ,
sir if my api has multiple route how do I deal with OPTIONS method problem ,
Should I need to create options route for each post request 
as You created for login in your application 
is that right solution for this problem ?????
Thanks

**Comment by irejwanul** [comment]
If we make post request with JavaScript Object as Param, then the 'Content-Type' requires to be set as 'application/json'. The JavaScript object can be passed on request body without any kind of conversion to string or anything. Such in my case, the server side technology is ASP.Net Core, and I can make angular http post request as follow
`updateProduct(product: IProductUpdateModel): Observable<IMyResponse> {
        /**
         * This check is required, cause in Server JS 'undefined', considers as string input rather than null string
         */
        if(product) {
            const xsrf: string = getStoreXsrf();
            let headers: HttpHeaders = new HttpHeaders()
                                        .set('RequestVerificationToken', xsrf)
                                        .set('Content-Type', 'application/json;charset=utf-8;');
            const postUrl: string = 'api/apiProduct/Update';

            return this.httpClient.post<IMyResponse>(postUrl, product, { headers: headers });
        } else {
            return this.httpClient.post<IMyResponse>('', '');
        }
    }`

**Comment by miguelleitevieira** [comment]
Hi all,

`
var body = 'email='+username+'&password='+password;

var config = {
     headers: new HttpHeaders({
          'Content-Type':  'application/x-www-form-urlencoded; charset=UTF-8',
     })
};

var payload = {
     params: {
          email: username,
          password: password
     },
     observe: 'response'
};

return this.httpClient.post<any>(url, payload, config)
`
I'm trying to put this working. I need to pass the body and the observe notation to get the full response, but I don't know how to do that. Someone can help?

**Comment by irejwanul** [comment]
Hi miguelleitevieira,
  What is your server side technology? What is your exact problem?

**Comment by sayjavajava** [comment]
@irejwanul would you like to help me 

I am using front end side angular 5
server side spring boot (Oauth2)

code 
```
`  login(usercreds){
    let headers = new HttpHeaders();
    headers = headers.append("Authorization", "Basic " + btoa("waqas:waqas-secret"));
    headers = headers.append("Content-Type", 'application/x-www-form-urlencoded;charset=UTF-8');

     var creds = 'usernamename=' + usercreds.username + '&password=' + usercreds.password + 
       "&grant_type=password"+ "&credentials=true" ;

return    this.httplclient.post('/api/oauth/token',creds, {headers: headers}).subscribe(response => {
          console.log(response);
    }, err => {
       console.log("User authentication failed!");
    });
          
  }


```

NETWORK:

![headers](https://user-images.githubusercontent.com/17933069/37552127-a16f3922-29d0-11e8-8a6d-5ee6d1d02960.png)


Error: Written [error="invalid_grant", error_description="Bad credentials"] as "application/json" using [org.springframework.http.converter.json.MappingJackson2HttpMessageConverter@290aeb20]

**Comment by irejwanul** [comment]
Hi @sayjavajava,
Can you provide your server side code as well.

**Comment by sayjavajava** [comment]
@irejwanul thanks i have solved issue . Now i am thinking about handling access_token and refresh_teoken , how should i manage them ? should i have store them in localstorage ? have you used them in client side before? actually i am doing 1st time .

**Comment by mohamedHassanKa** [comment]
@TitaneBoy i am having the same probleme as you i am stuck in OPTIONS http://localhost:3000/user/login net::ERR_CONNECTION_REFUSED   here is some photo of the server and the front end side 
![1](https://user-images.githubusercontent.com/33184775/38355982-b7d988be-38be-11e8-9a5d-0958e2a08745.PNG)
![2](https://user-images.githubusercontent.com/33184775/38355983-b805a2d2-38be-11e8-8650-13319d9cdecd.PNG)
and can you tell me what do you use to debug like that in  node js and thnx

**Comment by jmesa-sistel** [comment]
In the example he is using 'application/x-www-form-urlencoded' so we have to use form encoding.
I have created my own HttpFormEncodingCodec
this is what I am doing to encode/decode properly.

```
import { HttpParameterCodec } from '@angular/common/http';

export class HttpFormEncodingCodec implements HttpParameterCodec {
    encodeKey(k: string): string { return encodeURIComponent(k).replace(/%20/g, '+'); }

    encodeValue(v: string): string { return encodeURIComponent(v).replace(/%20/g, '+'); }

    decodeKey(k: string): string { return decodeURIComponent(k.replace(/\+/g, ' ')); }

    decodeValue(v: string) { return decodeURIComponent(v.replace(/\+/g, ' ')); }
}

.....................
doPost(description, name) {
    headers: HttpHeaders = new HttpHeaders()
        .append('Content-Type', 'application/x-www-form-urlencoded; charset=UTF-8');
    const body = new HttpParams({ encoder: new HttpFormEncodingCodec() })
        .append('description', description)
        .append('name', name)
        .toString();

    return this.http.post<any>(this.apiUrl, body, { headers: headers });
}
```

If you are going to use the HttpFormEncodingCodec several times, you should create it only once in the class constructor to reuse it.
Hope it helps

**Comment by sispor** [comment]
I don't know it's bad, but you can use:

Client:

    register(user: User): Observable<any>{

        let body: string = JSON.stringify(user);
        
        let params = new HttpParams().set('params',body);
        console.log(params);
        let headers = new HttpHeaders().set('Content-type','application/json');

        return this._http.post(this.url+'register',
        {"body":body}, {
            headers: headers,
            params
        })
        .pipe(
            catchError(error => {
                return throwError("something went wrong!");
            }
        ));
    }

Host:

function saveUser(req,res){
    var params = req;
    console.log(params);
    var user = new User();
    if(params.name && params.surname && params.nick 
        && params.email && params.password){
        user.name= params.name;
        user.surname = params.surname;
        user.nick = params.nick;
        user.email= params.email;
        user.role = "ROLE_USER";
        user.image =null;
        
        User.findOne({$or: [
            {email: user.email.toLowerCase()},
            {nick: user.nick.toLowerCase()}
        ]}).exec((err, users) => {//exe para ejecutar el calba
        if (err) return res.status(500).send({message: "error en peticion de usuarios"})
        if (users && users.length >= 1){
            return res.status(200).send({message:"este usuario ya existe"})
        }else{
            bcrypt.hash(params.password,null, null, (err, hash) =>{
                user.password = hash;
               user.save((err, userStored)=>{
                   if(err) return res.status(500).send({message: "error al guardar el usuario"});
                   if(userStored){
                       res.status(200).send({user: userStored});
                   }else{
                       res.status(404).send({message: "no se ha registrado el usuario"});
                   }
               });
            });
        }
    });

        
    }
    else{
        res.status(200).send({
            message: "rellena todos los campos"
        });
    }
}
